engine/modules/opentelemetry/src/telegraf/nagios_converter.cc

```diff
@@ -0,0 +1,276 @@
+ * @brief
+ *
+ * @param fifos fifos indexed by metric_name such as check_icmp_critical_gt,
+ * check_icmp_state
+ * @return com::centreon::engine::commands::result
+ */
+bool otl_nagios_converter::_build_result_from_metrics(
+    metric_name_to_fifo& fifos,
+    commands::result& res) {
+  // first we search last state timestamp
+  uint64_t last_time = 0;
+
+  for (auto& metric_to_fifo : fifos) {
+    if (get_nagios_telegraf_suffix(metric_to_fifo.first) == "state") {
+      auto& fifo = metric_to_fifo.second.get_fifo();
+      if (!fifo.empty()) {
+        const auto& last_sample = *fifo.rbegin();
+        last_time = last_sample.get_nano_timestamp();
+        res.exit_code = last_sample.get_value();
+        metric_to_fifo.second.clean_oldest(last_time);
+      }
+      break;
+    }
+  }
+  if (!last_time) {
+    return false;
+  }
+  res.command_id = get_command_id();
+  res.exit_status = process::normal;
+  res.end_time = res.start_time = last_time / 1000000000;
+
+  // construct perfdata list by perfdata name
+  std::map<std::string, detail::perf_data> perfs;
+
+  for (auto& metric_to_fifo : fifos) {
+    std::string_view suffix = get_nagios_telegraf_suffix(metric_to_fifo.first);
+    const data_point_fifo::container& data_points =
+        metric_to_fifo.second.get_fifo();
+    // we scan all data points for that metric (example check_icmp_critical_gt
+    // can contain a data point for pl and another for rta)
+    auto data_pt_search = data_points.equal_range(last_time);
+    for (; data_pt_search.first != data_pt_search.second;
+         ++data_pt_search.first) {
+      const auto attributes = data_pt_search.first->get_data_point_attributes();
+      std::string perfdata_name;
+      std::string unit;
+      for (const auto& attrib : attributes) {
+        if (attrib.key() == "perfdata") {
+          perfdata_name = attrib.value().string_value();
+        } else if (attrib.key() == "unit") {
+          unit = attrib.value().string_value();
+        }
+        if (!perfdata_name.empty() && !unit.empty()) {
+          break;
+        }
+      }
+      if (!perfdata_name.empty()) {
+        detail::perf_data& to_fill = perfs[perfdata_name];
+        if (!unit.empty()) {
+          to_fill.unit = unit;
+        }
+        to_fill.fill_from_suffix(suffix, data_pt_search.first->get_value(),
+                                 _logger);
+      }
+    }
+    metric_to_fifo.second.clean_oldest(last_time);
+  }
+
+  data_point_fifo_container::clean_empty_fifos(fifos);
+
+  // then format all in a string with format:
+  // 'label'=value[UOM];[warn];[crit];[min];[max]
+  if (res.exit_code >= 0 && res.exit_code < 4) {
+    res.output = state_str[res.exit_code];
+  }
+  res.output.push_back('|');
+  for (const auto& perf : perfs) {
+    if (perf.second.val) {
+      res.output += perf.first;
+      res.output.push_back('=');
+      absl::StrAppend(&res.output, *perf.second.val);
+      res.output += perf.second.unit;
+      res.output.push_back(';');
```

Comment:
No way to do something like
```
res.output += fmt::format("{}={}{};", perf.first, *perf.second.val, perf.second.unit);
```
Or with the absl functions....

Reply:
And maybe we can improve a little more with min, max, and thresholds.

Reply:
done